src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -57,7 +57,12 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
 
   /** Reduce length to `n`, nulling out all dropped elements */
   private def reduceToSize(n: Int): Unit = {
-    RefArrayUtils.nullElems(array, n, size0)
+    val reducedSize = Math.min(n, size0)
+    val reducedLength = Math.max(initialElements.length, reducedSize)
+    if (array.length / 2  > reducedLength)
+      array = RefArrayUtils.ensureSize(array, reducedSize, reducedLength)
+    else
+      RefArrayUtils.nullElems(array, n, size0)
```

Comment:
`ensureSize` only shrinks when it's 4 times too large, but you're not nulling unless it's no more than 2 times too large.